source/grammar/qasm3Parser.g4

```diff
@@ -0,0 +1,361 @@
+
+classicalType
+    : singleDesignatorType designator
+    | noDesignatorType
+    | bitType designator?
+    | COMPLEX LBRACKET numericType RBRACKET
+    ;
+
+// numeric OpenQASM types
+numericType: singleDesignatorType designator;
+
+constantDeclaration: CONST Identifier equalsExpression;
+
+// if multiple variables declared at once, either none are assigned or all are assigned
+// prevents ambiguity w/ qubit arguments in subroutine calls
+singleDesignatorDeclaration: singleDesignatorType designator Identifier equalsExpression?;
+
+noDesignatorDeclaration: noDesignatorType Identifier equalsExpression?;
+
+bitDeclaration: ( CREG Identifier designator? | BIT designator? Identifier ) equalsExpression?;
+complexDeclaration: COMPLEX LBRACKET numericType RBRACKET Identifier equalsExpression?;
+
+classicalDeclaration
+    : singleDesignatorDeclaration
+    | noDesignatorDeclaration
+    | bitDeclaration
+    | complexDeclaration
+    ;
+
+classicalTypeList: classicalType (COMMA classicalType)*;
+classicalArgument
+    :
+    (
+        singleDesignatorType designator |
+        noDesignatorType
+    ) Identifier
+    | CREG Identifier designator?
+    | BIT designator? Identifier
+    | COMPLEX LBRACKET numericType RBRACKET Identifier
+    ;
+
+classicalArgumentList: classicalArgument (COMMA classicalArgument)*;
+
+anyTypeArgument: classicalArgument | quantumArgument;
+anyTypeArgumentList: anyTypeArgument (COMMA anyTypeArgument)*;
+
+/** Aliasing **/
+aliasStatement: LET Identifier EQUALS indexIdentifier SEMICOLON;
+
+/** Register Concatenation and Slicing **/
+
+indexIdentifier
+    : Identifier rangeDefinition
+    | Identifier ( LBRACKET expressionList RBRACKET )?
+    | indexIdentifier DOUBLE_PIPE indexIdentifier
+    ;
+
+indexIdentifierList
+    : indexIdentifier ( COMMA indexIdentifier )*
+    ;
+
+rangeDefinition
+    : LBRACKET expression? COLON expression? ( COLON expression )? RBRACKET
+    ;
+
+/*** Gates and Built-in Quantum Instructions ***/
+
+quantumGateDefinition
+    : GATE quantumGateSignature quantumBlock
+    ;
+
+quantumGateSignature
+    : quantumGateName ( LPAREN identifierList? RPAREN )? identifierList
+    ;
+
+quantumGateName: U_ | CX | Identifier;
+quantumBlock
+    : LBRACE ( quantumStatement | quantumLoop )* RBRACE
+    ;
+
+// loops containing only quantum statements allowed in gates
+quantumLoop
+    : loopSignature quantumLoopBlock
+    ;
+
+quantumLoopBlock
+    : quantumStatement
+    | LBRACE quantumStatement* RBRACE
+    ;
+
+quantumStatement
+    : quantumInstruction SEMICOLON
+    | timingStatement
+    ;
+
+quantumInstruction
+    : quantumGateCall
+    | quantumPhase
+    | quantumMeasurement
+    | quantumReset
+    | quantumBarrier
+    ;
+
+quantumPhase
+    : quantumGateModifier* GPHASE LPAREN expression RPAREN indexIdentifierList?
+    ;
+
+quantumReset
+    : RESET indexIdentifierList
+    ;
+
+quantumMeasurement
+    : MEASURE indexIdentifier
+    ;
+
+quantumMeasurementAssignment
```

Comment:
On line 149, in `quantumInstruction` rule, we have `quantumMeasurement` but it cannot be a `quantumMeasurementAssignment`. Is there a reason to not include assignment there? (I am also wondering what's the expected compiler behavior if the measurement results are not stored)

If indeed we don't want to include "assignment" in `quantumInstruction` for some reason, I suggest we can make the `(ARROW indexIdentifier)?` part on line 170 not optional. It will not change the grammar. Because in `statement`, we have both `quantumStatement` (thus `quantumInstruction`) and `assignmentStatement` (thus `quantumMeasurementAssignement`. This will bring more clarity to the parser.

Reply:
I didn't change any of these parser rules in the split (as far as I remember!).  In general, I'd like to keep it that way, so this PR is just a (mostly) clean application of the split, then we can do a bunch of follow-ups to fix all these little grammatical niggles.

On the actual grammar issue, though, yeah, I agree.